### PR TITLE
chartsettings-support: support for adding chart settings

### PR DESCRIPTION
**Pull request by cielowu**

### CODE REVIEW

File: src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java [hunk 1]
```diff
@@ -119,6 +127,18 @@ public ChartSlugBuilderImpl setBase(int base) {
     return this;
   }
 
+  @Override
+  public ChartSlugBuilder setChartSettings(String chartSettings) {
```
Comment (kentwang929) [inline src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:131]: why not just set this to ChartSettings?

**Review by kentwang929 — APPROVED** [review]
lgtm, one comment.

File: src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java (hunk 1 above)
Comment (kentwang929) [reply src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:131]: or compatible for both input

Comment (cielowu) [reply src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:131]: I was concerned about exposing ChartSettings at first, as all other structured objects (Chart, ChartSlug) are considered internal / `package private`.
But maybe we should not have a fat builder and expose all of the objects in the first place?

Comment (cielowu) [reply src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:131]: I think we want to make the user only create the slug through the builder interface.
If we expose ChartSettings and other classes, we won't be able to restrict this and the user is able to build a slug in a way we might not expect? Exposing chart settings only works, but is just not consistent with others.

Comment (kentwang929) [reply src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:131]: let's chat

File: src/main/java/com/wavefront/slug/chart/ChartSlugBuilder.java [hunk 2]
```diff
@@ -105,6 +105,30 @@
...
+  /**
+   * Sets the chart's settings from a JSON serialized string.
+   *
+   * @param chartSettings chart settings string
+   * @return The builder.
+   */
+  ChartSlugBuilder setChartSettings(String chartSettings);
+
+  /**
+   * Sets the chart's settings from a JSON serialized string.
```
Comment (kentwang929) [inline src/main/java/com/wavefront/slug/chart/ChartSlugBuilder.java:117]: update comment

File: src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java [hunk 3]
```diff
@@ -114,11 +122,28 @@ public ChartSlugBuilderImpl setUnits(String units) {
...
+    this.chartSettings = parseJSON(chartSettings, ChartSettings.class);
+    return this;
+  }
+
+  @Override
+  public ChartSlugBuilder setChartSettings(ChartSettings chartSettings) {
+    this.chartSettings = chartSettings;
+    return null;
```
Comment (kentwang929) [inline src/main/java/com/wavefront/slug/chart/ChartSlugBuilderImpl.java:138]: return this

**Review by kentwang929 — CHANGES_REQUESTED** [review]
fix the comments and the builder return, then its good

**Review by kentwang929 — APPROVED** [review]

**Pull request merged**